src/parser.cpp

```diff
@@ -26,7 +27,7 @@ namespace ada::parser {
    * to_ascii does not expect the input to be percent decoded. This is
    * mostly used to conform with the test suite.
    */
-  std::optional<std::string> to_ascii(const std::string_view plain, const bool be_strict, size_t first_percent) {
+  bool to_ascii(std::optional<std::string>& out, const std::string_view plain, const bool be_strict, size_t first_percent) {
```

Comment:
This does not look like a clear gain. The new code looks more complicated.

Reply:
This is actually 200ns slower:

`to_ascii(std::optional<std::string>& out)`:

```
BasicBench_AdaURL            4561 ns         4560 ns       154164 time/byte=5.64395ns time/url=414.574ns url/s=2.41212M/s
```

`std::optional<std::string>& to_ascii()`:

```
BasicBench_AdaURL       4780 ns         4778 ns       146452 time/byte=5.91315ns time/url=434.348ns url/s=2.3023M/s
```

Reply:
Is it correct though?

Reply:
I can open a different pull request and you can run locally

Reply:
@anonrig Yes, the current PR fails the tests.

Reply:
I fixed the tests, and created another pull request to this branch with your suggested changes: https://github.com/ada-url/ada/pull/62